feat: rule improvements

Laurin-W opened the pull request:
- Change authentication method rule result description to say token_endpoint_auth_method with client_secret_basic is not supported for Client Identifier Documents (instead of not supported by Solid OIDC)

- Add no localhost Client ID rule

- Add whitespace-only client name rule result

Laurin-W committed:
feat: add whitespace-only client name rule

Laurin-W committed:
chore: add isUriLocalhost helper function

Laurin-W committed:
feat: add no localhost Client ID rule

Laurin-W committed:
feat: improve authentication method rule result description

ThisIsMissEm reviewed (commented):
This looks okay, but be careful with combining multiple unrelated changes into one PR

Laurin-W commented on src/lib/validationRules/decentClientName/decentClientName.test.ts:
```diff
@@ -42,6 +42,18 @@ describe("well-formed client name check", () => {
+  it("warns on whitespace client name", async () => {
+    const resultsForLongClientName = await decentClientName.check({
+      document: {
+        client_name: " \n\t\v\r\f",
```
Do you think, I should just leave them out here?

Laurin-W commented on src/lib/helperFunctions.ts:
```diff
@@ -45,3 +45,24 @@ export const statusToNumber = (
+  if (hostname === "localhost" || hostname === "[::1]") {
+    return true;
+  }
+  if (hostname.startsWith("127.0.0.") && hostname.length <= 11) {
```
Why not IPv6? `::1` is the only valid ip6 loopback address as far as I'm aware and since it's URI hostnames, the address surrounded by the square braces or am I getting you wrong?

Laurin-W replied:
Changed test to support the whole range with 11667c0

Laurin-W committed:
adopt feedback

- client name rule result title
- client name rule test data with spaces-only whitespace
- add support for the whole 127.0.0.0/8 range in isHostnameLocal

Laurin-W commented on src/lib/validationRules/decentClientName/decentClientName.ts:
```diff
@@ -56,6 +56,18 @@ const decentClientName: ValidationRule = {
+      return [
+        {
+          status: "warning",
+          title: "Client Name is whitespace only",
```
changed with 11667c0

ThisIsMissEm commented on src/lib/validationRules/noLocalhostClientId/noLocalhostClientId.ts:
```diff
@@ -0,0 +1,49 @@
+          status: "warning",
+          title: "Localhost Client Identifier",
+          description:
+            "Client Identifier URIs must be dereferenced by the OIDC Provider and are therefore only useful for development when your client and OP run on the same machine.",
```
```suggestion
            "Client Identifier URIs must be dereferenced by the OIDC Provider and are therefore only useful for development when your client application and OIDC Provider run on the same machine.",
```

This description kind of duplicates the above description, I'd suggest setting one of them as empty or deduplicating what they say.

Laurin-W replied:
changed with https://github.com/inrupt/client-identifier-helper/pull/133/commits/e6c5bca02f136cd11a3da147a50f5cfa97750de4

ThisIsMissEm commented on src/lib/validationRules/noLocalhostClientId/noLocalhostClientId.ts:
```diff
@@ -0,0 +1,49 @@
+const noLocalhostClientId: ValidationRule = {
+  rule: {
+    type: "local",
+    name: "Localhost Client IDs in development only",
```
```suggestion
    name: "Client Identifier uses localhost for client_id",
```

Laurin-W replied:
changed with https://github.com/inrupt/client-identifier-helper/pull/133/commits/e6c5bca02f136cd11a3da147a50f5cfa97750de4

ThisIsMissEm commented on src/lib/validationRules/rightAuthenticationMethod/rightAuthenticationMethod.ts:
```diff
@@ -25,7 +25,7 @@ const rightAuthenticationMethod: ValidationRule = {
     name: "Field `token_endpoint_auth_method` must be `none`.",
     description:
-      "Solid OIDC connect does only support `token_endpoint_auth_method` set to `none`. This must be made explicit to comply with the OIDC specification.",
+      "Authentication with Client Identifier Documents is only supported with `token_endpoint_auth_method` set to `none`. This must be made explicit to comply with the Solid OIDC specification.",
```
You might want to add a note that not all Solid servers are strict about this compliance point — ESS currently isn't.

Laurin-W replied:
changed with https://github.com/inrupt/client-identifier-helper/pull/133/commits/25a2e821818647566be8a36114feff15d18f5fd2

ThisIsMissEm commented on src/lib/helperFunctions.ts:
```diff
@@ -45,3 +45,25 @@ export const statusToNumber = (
+  // Test 127.0.0.0/8 range.
+  const blocks = hostname.split(".");
+  if (blocks.length === 4 && blocks[0] === "127") {
+    return blocks
+      .slice(1)
+      .every((block) => Number(block) >= 0 && Number(block) < 256);
```
This can just `return true` — because 127.500.500.20` is not a valid IP, or alternatively:

```
if (LOCALHOST_IP_REGEX.test(hostname))
```

where `LOCALHOST_IP_REGEX` == `/^127.\d{1,3}.\d{1,3}.\d{1,3}$/`

Laurin-W replied:
Mhh, the reason why I didn't go this way was because Firefox will parse `new URL("https://127.999.999.999")` (chromium doesn't) and hence `isUriLocalhost(uri: string)` would return true for that URI.

ThisIsMissEm replied:
You could still use a regex here instead of split/every

Laurin-W replied:
You mean like this?
```typescript
  const LOCALHOST_IP_REGEX = /^127.(\d{1,3}).(\d{1,3}).(\d{1,3})$/;

  // Test 127.0.0.0/8 range.
  if (LOCALHOST_IP_REGEX.test(hostname)) {
    LOCALHOST_IP_REGEX.exec(hostname)
      ?.slice(1)
      .every((block) => Number(block) >= 0 && Number(block) < 256);
  }
  ```

ThisIsMissEm replied:
``` const LOCALHOST_IP_REGEX = /^127.(\d{1,3}).(\d{1,3}).(\d{1,3})$/;

  // Test 127.0.0.0/8 range.
  if (LOCALHOST_IP_REGEX.test(hostname)) {
    return true;
  }
```

Laurin-W replied:
done with 5911a5375e455439c2e0c07bb1c6fd1742ab47d0

ThisIsMissEm commented on src/lib/helperFunctions.ts:
```diff
@@ -45,3 +45,25 @@ export const statusToNumber = (
+  return false;
+};
+
+export const isUriLocalhost = (uri: string) => {
```
Careful on return types here, as you've effectively got: `isUriLocalhost(uri: string): boolean | undefined` which would be what's considered a "tri-state boolean", which is totally a misnomer. if you encounter an error, it's probably safe to return `false`

Laurin-W replied:
changed with https://github.com/inrupt/client-identifier-helper/pull/133/commits/57f3170d9aded65e9e2ac540f765b91b0bdd5be8

ThisIsMissEm replied:
https://en.wikipedia.org/wiki/Boolean_data_type — _In [computer science](https://en.wikipedia.org/wiki/Computer_science), the Boolean (sometimes shortened to Bool) is a [data type](https://en.wikipedia.org/wiki/Data_type) that has one of two possible values (usually denoted true and false)_

So a value which is true / false / undefined really isn't a boolean, it's some third thing that doesn't generally make sense.

ThisIsMissEm commented on src/lib/validationRules/decentClientName/decentClientName.ts:
```diff
@@ -56,6 +56,18 @@ const decentClientName: ValidationRule = {
+          title: "Invalid Client Name",
+          description:
+            "The Client Name consists of whitespace only. It should be set for authentication providers to display the name to the end-user.",
+          affectedFields: [{ fieldName: "client_name", fieldValue: undefined }],
```
```suggestion
          affectedFields: [{
            fieldName: "client_name",
            fieldValue: context.document.client_name,
          }],
```

Laurin-W replied:
fixed with https://github.com/inrupt/client-identifier-helper/pull/133/commits/3947ac0d7ee6f6f7ac174ea954d6e1cc64281ae8

ThisIsMissEm commented on src/lib/validationRules/noLocalhostClientId/noLocalhostClientId.test.ts:
```diff
@@ -0,0 +1,50 @@
+
+/* eslint-disable-next-line no-shadow */
+import { describe, expect, it } from "@jest/globals";
+import decentClientName from "./noLocalhostClientId";
```
```suggestion
import noLocalhostClientId from "./noLocalhostClientId";
```

And then change all the tests usage.

Laurin-W replied:
addressed with https://github.com/inrupt/client-identifier-helper/pull/133/commits/5978b5b1f91a988d375fc7ce1f0ae52e37ef12c9

Laurin-W committed:
improve noLocalhostClientId texts

Laurin-W committed:
improve rightAuthenticationMethod texts

Laurin-W committed:
adjust isUriLocalHost helper function

Laurin-W committed:
add affectedFields fieldValue for rule results

Laurin-W committed:
correct imported test name for noLocalHostClientId

ThisIsMissEm commented on src/lib/validationRules/noMixedRedirectUris/noMixedRedirectUrls.ts:
```diff
@@ -18,6 +18,7 @@
 // OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 // SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 //
+import { isHostnameLocal, isUriLocalhost } from "../../helperFunctions";
```
Please fix the lint warning

Laurin-W replied:
done with 81b6eabeda8b46519ee57062a77cb0f5613189e3

ThisIsMissEm commented on src/lib/validationRules/noLocalhostClientId/noLocalhostClientId.ts:
```diff
@@ -0,0 +1,50 @@
+import { isUriLocalhost } from "../../helperFunctions";
+import { ValidationContext, ValidationRule } from "../../types";
+
+const descriptionMessage =
```
Note quite want I meant by de-duplicate, but fine

Laurin-W replied:
ahh sorry, now I get what you meant  😅

Laurin-W committed:
isHostnameLocal with regex

Laurin-W committed:
remove linting warning

Laurin-W committed:
change noLocalhostClientId message
